packages/slonik/src/service.ts

```diff
@@ -180,9 +206,52 @@ abstract class BaseService<
     return this.factory.table;
   }
 
+  protected async preProcess<D>(
+    action: string,
+    data?: D,
+  ): Promise<D | undefined> {
+    const hookName = `pre${action.charAt(0).toUpperCase()}${action.slice(1)}`;
+
+    const preHookFunction = (this as Record<string, unknown>)[hookName];
+
+    if (typeof preHookFunction === "function") {
+      const preHook = preHookFunction as (data?: D) => Promise<D | undefined>;
+
+      const processedData = await preHook(data);
+
+      // FIXME only performing a shallow runtime check, using JavaScript's typeof,
+      // which is limited and not reliable for complex data types like arrays, objects, classes, etc.
+      if (typeof processedData === typeof data) {
```

Comment:
Since javascript is weakly typed, Im not sure we can really enforce the type check at runtime. Typescript will enforce this at compile time (won't it?), and maybe this is good enough.